Give PATH input more functionality

reSoley opened the pull request:
PATH now supports paths to single files within the application to be downloaded. PATH also supports globs at the base of the file path. Multiple paths can be given as input.

reSoley committed:
Input PATH now supports paths to files.

reSoley committed:
Remove "app-download" dir for single file or dir.

reSoley committed:
Only download endpoint of input PATH.

reSoley committed:
Input PATH now supports multiple paths.

reSoley committed:
Input PATH now supports globs at the base.

reSoley committed:
Added complete documentation to main.go.

whereswaldon commented on README.md:
```diff
@@ -28,17 +28,25 @@
+cf download APP_NAME [PATH...] [--overwrite] [--file] [--verbose] [--omit omitted_path] [-i instance]
 
-The downloaded app files will put put in a new directory "APP_NAME-download" that's created within your working directory.
+The downloaded app files will be put in a new directory "APP_NAME" that's created within your working directory.
```
So, after your changes the downloaded app's directory will not be suffixed with `-download`?

reSoley replied:
If the entire application is being downloaded, yes. If a specific file or directory is being downloaded it will be placed in the current working directory.

whereswaldon replied:
Okay, huh. That really isn't how the comment above reads to me.

whereswaldon commented on main.go:
```diff
@@ -42,6 +42,14 @@ type flagVal struct {
+type pathVal struct {
+	RootWorkingDirectoryLocal  string
+	RootWorkingDirectoryServer string
+	StartingPathServer         string
```
So pathVal holds the working directory for both the local and server hosts, as well as the "starting path" for the server. Is that the starting path of an individual file, or something else?

reSoley replied:
Looking at how RootWorkingDirectoryServer is used, I realize that it is redundant and unnecessary. It is only passed to NewDownloader and within that initializer it is transformed into the same path as StartingPathServer.

whereswaldon replied:
Okay, so how do you plan to handle that? Refactor to remove it? What was your intention when you created it?

reSoley replied:
Yup, a refactor to get rid of it and get `Download` to just take `StartingPathServer` is the plan. `GetDirectoryContext` returned three paths when I began work on this, so I figured the last person to update it had a reason.

whereswaldon replied:
I think @jlstack did that, maybe for a reason. Hopefully he might chime in here now that I've mentioned him.

whereswaldon commented on main.go:
```diff
@@ -88,108 +96,175 @@ func (c *DownloadPlugin) Run(cliConnection plugin.CliConnection, args []string)
-	rootWorkingDirectoryLocal := filepath.FromSlash(rootWorkingDirectoryServer)
+	pathVals := GetDirectoryContext(workingDir, paths, flagVals.File_flag)
+	for _, v := range pathVals {
+		v.RootWorkingDirectoryLocal = filepath.FromSlash(v.RootWorkingDirectoryLocal)
```
If I understand this line correctly, it postprocesses all of the RootWorkingDirectoryLocal values to make them OS-agnostic. Is there a reason that this isn't done inside of `GetDirectoryContext`? It seems like you probably don't want your helper functions emitting OS-unsafe paths, especially given that `GetDirectoryContext` is visible outside of the module.

reSoley replied:
Good point. There's no specific reason.

jlstack replied:
The rootWorkingDirectoryServer and rootWorkingDirectoryLocal variables were created to try to prevent confusion. As @whereswaldon mentioned, the local variable needed to be OS-agnostic while the server variable was for a Linux server. @reSoley is correct though when he says it is merely passed into another function and makes no discernible difference.

whereswaldon commented:
Kind of a general question: Are there any golang formatting guidelines for line length? I know that the compiler doesn't enforce any, but I do find it harder to read code that gets as long as this. It would be nice to truncate lines at the 120-char point, if not at 80. Just a thought though.

whereswaldon commented on main.go:
```diff
@@ -88,108 +96,175 @@ func (c *DownloadPlugin) Run(cliConnection plugin.CliConnection, args []string)
 	}
 
-	flagVals := ParseFlags(args)
+	cmdExec := cmd_exec.NewCmdExec()
```
So this `cmdExec` is shared by all of the downloads?

reSoley replied:
Yes. The `dir_parser` on line 136 can also be shared by all the downloads so I'll move that initialization out of the for loop as well.

whereswaldon replied:
Nice catch! Sounds like a good idea.

whereswaldon commented on main.go:
```diff
@@ -88,108 +96,175 @@ func (c *DownloadPlugin) Run(cliConnection plugin.CliConnection, args []string)
+
+			// start download of directory
+			wg.Add(1)
+			dloader.Download(files, dirs, v.StartingPathServer, v.RootWorkingDirectoryLocal, filterList)
```
Why doesn't `dloader.Download` need to accept the waitgroup pointer like `dloader.DownloadFile`? You still add one to the waitgroup, so how does that download goroutine get associated with the wait group?

reSoley replied:
The `dloader` object stores a pointer to the wait group itself when initialized. It actually may not be necessary to pass it to `DownloadFile` either since the `dloader` already has it.

whereswaldon replied:
Ah, that makes sense. In that case, does it make sense to pass one to `dloader.DownloadFile`? Maybe that's also unnecessary?

whereswaldon replied:
This concern has been addressed by  6717206

whereswaldon commented on main.go:
```diff
@@ -88,108 +96,175 @@ func (c *DownloadPlugin) Run(cliConnection plugin.CliConnection, args []string)
-	// parse the directory
-	files, dirs := parser.ExecParseDir(startingPathServer)
+		parser = dir_parser.NewParser(cmdExec, appName, flagVals.Instance_flag, onWindows, flagVals.Verbose_flag)
+		dloader = downloader.NewDownloader(cmdExec, &wg, appName, flagVals.Instance_flag, v.RootWorkingDirectoryServer, flagVals.Verbose_flag, onWindows)
```
Where does `wg` get initialized? I see it declared as a var, but never allocated. Does that happen implicitly somewhere?

reSoley replied:
This happens when `wg.Add(1)` is called before the call to `Download` on line 154.

whereswaldon replied:
Indeed it is. It's very strange that calling `wg.add(n)` initializes it, but that's how the [usage example](https://golang.org/pkg/sync/#WaitGroup) does it too.

whereswaldon commented on main.go:
```diff
@@ -88,108 +96,175 @@ func (c *DownloadPlugin) Run(cliConnection plugin.CliConnection, args []string)
-		quit <- 0
+		// stop console writer
+		if flagVals.Verbose_flag == false {
+			quit <- 0
```
If each download is a goroutine within a wait group, why do you need to send a kill down this quit channel? Shouldn't all of those goroutines be done after line 165?

reSoley replied:
This kill is sent to the console writer, which is still running at the time this happens.

whereswaldon commented on main.go:
```diff
@@ -88,108 +96,175 @@ func (c *DownloadPlugin) Run(cliConnection plugin.CliConnection, args []string)
+	var pathVals []pathVal
 
-	// append path if provided as arguement
+	rootWorkingDir := workingDir + "/"
```
Why not use [`os.PathSeparator`](https://golang.org/pkg/os/#PathSeparator) as a suffix instead of `/`? Might also be applicable to line 200

whereswaldon commented on main.go:
```diff
@@ -88,108 +96,175 @@ func (c *DownloadPlugin) Run(cliConnection plugin.CliConnection, args []string)
+
+	if len(paths) == 0 {
+		// create appName directory if downloading whole app
+		addPathVals := pathVal{
```
Maybe don't even worry about `os.PathSeparator`, but call `filepath.FromSlash` on the `RootWorkingDirectoryLocal` field after you concatenate those strings?

whereswaldon commented on main.go:
```diff
@@ -88,108 +96,175 @@ func (c *DownloadPlugin) Run(cliConnection plugin.CliConnection, args []string)
+			}
+
+			addPathVals := pathVal{
+				RootWorkingDirectoryLocal:  localPath + filepath.Base(v),
```
Also do a `filepath.FromSlash` on this `RootWorkingDirectoryLocal`

whereswaldon commented on main.go:
```diff
@@ -210,22 +285,23 @@ func ParseFlags(args []string) flagVal {
 /*
-*	consoleWriter prints the current number of files downloaded. It is polled every 350 milleseconds
-* 	disabled if using verbose flag.
+*	This funciton prints the current number of files downloaded. It is polled every 350 milleseconds
```
The word `function` is misspelled. Might as well fix it.

whereswaldon commented on main.go:
```diff
@@ -309,6 +394,41 @@ func Exists(path string) bool {
+			dir := filepath.Dir(v)
+			out, _ := cmdExec.GetFile(appName, dir, instance)
+			body := strings.Split(string(out), "\n")
+			body = body[3 : len(body)-2]
```
I think it would probably be worthwhile to add a comment on this line that explains why you sliced this array this way.

reSoley replied:
Good point. It isn't exactly intuitive.

whereswaldon commented on main.go:
```diff
@@ -309,6 +394,41 @@ func Exists(path string) bool {
+				match, err := filepath.Match(filepath.Base(v), strings.TrimSuffix(cur, "/"))
+				check(err, "")
+				// append path to return list if glob pattern matches
+				if match && !strings.HasPrefix(cur, ".") {
```
Does this line ignore hidden files?

reSoley replied:
Yes, because the `dloader` ignores them as well when downloading directories or the entire application.

whereswaldon replied:
Is there an option to turn that off? I'm surprised. I can think of many times when I'd want to download hidden config files and the like. It seems like a feature that should exist.

reSoley replied:
`cf files` on a hidden file fails to get the contents of it, so it can't be copied to a local file.

whereswaldon commented on main_test.go:
```diff
@@ -15,22 +15,37 @@ import (
 
 		Context("Check if overWrite flag works", func() {
 			It("Should set the overwrite_flag", func() {
+				args = make([]string, 4)
```
I totally understand why you allocated the arrays this way, but I think you have a better option. I found [this](https://blog.golang.org/go-slices-usage-and-internals) in the official docs:

> Or, you can have the compiler count the array elements for you: `b := [...]string{"Penn", "Teller"}`

I think it would make the test file a lot shorter, which is probably a good thing for long-term maintainability.

reSoley replied:
I think I tried that and encountered an issue, but I'll look into it again.

whereswaldon replied:
Maybe you can't because `args` is defined earlier, but uninitialized? I could maybe see it happening because we didn't do a `:=`-type assignment.

reSoley committed:
Remove redundant var from main.go & download.go.

As per whereswaldon's comments on pull request
8, RootWorkingDirectoryServer was removed as it
is only used to calculate a path that was
already stored.

reSoley committed:
Don't pass wg to DownloadFile, fix download tests.

As per discussion on pull request 8, WaitGroup
wg is no longer passed to DownloadFile as the
downloader struct keeps a pointer to it. Also
fixed download tests broken in previous commit.

reSoley committed:
Refactor main_test.go to improve readability.

As per discussion with whereswaldon on pull
request 8, the argument arrays have been
formatted more concisely.

reSoley committed:
Fix tests on Windows.

whereswaldon commented on downloader/download.go:
```diff
@@ -77,9 +75,7 @@ func (d *downloader) Download(files, dirs []string, readPath, writePath string,
-
-		if filter.CheckToFilter(filePath, filterList) {
+		if filter.CheckToFilter(fileRPath, filterList) {
 			continue
```
I might write an inline comment explaining that this continue statement skips the waitgroup add and goroutine creation. That's purely stylistic though, so I won't be offended if it isn't worth it to you.

whereswaldon commented on downloader/download.go:
```diff
@@ -91,9 +87,8 @@ func (d *downloader) Download(files, dirs []string, readPath, writePath string,
 
-		if filter.CheckToFilter(dirPath, filterList) {
+		if filter.CheckToFilter(dirRPath, filterList) {
 			continue
```
Duplicate: I might write an inline comment explaining that this continue statement skips the waitgroup add and goroutine creation. That's purely stylistic though, so I won't be offended if it isn't worth it to you.

whereswaldon commented on downloader/download.go:
```diff
@@ -91,9 +87,8 @@ func (d *downloader) Download(files, dirs []string, readPath, writePath string,
 	for _, val := range dirs {
 		dirWPath := writePath + filepath.FromSlash(val)
 		dirRPath := readPath + val
```
I like this idea of thinking in terms of Read and Write paths. I think it's mentally simpler that thinking in terms of server and local paths. That's a great convention; well done.

whereswaldon commented on main.go:
```diff
@@ -88,108 +94,174 @@ func (c *DownloadPlugin) Run(cliConnection plugin.CliConnection, args []string)
+	var pathVals []pathVal
 
-	// append path if provided as arguement
+	//rootWorkingDir := workingDir + "/"
```
Can we take out this old code?

whereswaldon commented on main.go:
```diff
@@ -199,10 +199,11 @@ func GetDirectoryContext(workingDir string, paths []string, isFile bool) []pathV
 			StartingPathServer:        startingPath,
 		}
 
+		addPathVals.RootWorkingDirectoryLocal = filepath.FromSlash(addPathVals.RootWorkingDirectoryLocal)
```
You **really** shouldn't have to do this. I understand that you tried and encountered problems the other way, but it baffles me. What gave you the hint to try doing this in the first place?

reSoley committed:
Update CHANGELOG.md
